ang/crmCaseType.js

```diff
@@ -460,31 +479,55 @@
       activity.default_assignee_contact = null;
     };
 
-    // TODO roleName passed to addRole is a misnomer, its passed as the
-    // label HOWEVER it should be saved to xml as the name see
-    // https://lab.civicrm.org/dev/core/issues/774
+    // Note that roleName passed to addRole is a misnomer, it's passed as the
+    // label, but we want to pass on as the internal name.
+    // Additionally, roleName is backwards from anything we have defined in
+    // $scope.caseType.definition.caseRoles, so we need to do some work to
+    // tell if it's already in the list, but it's work we have to do anyway
+    // to pass on the internal name.
 
     /// Add a new role
     $scope.addRole = function(roles, roleName) {
+      var matchingRole;
+      // does the label we've been given match up to any relationship type, based on id (which is label)
+      var matchingRoles = _.filter($scope.relationshipTypeOptions, {id: roleName});
+      if (matchingRoles.length) {
+        matchingRole = matchingRoles.shift();
+      }
+      // if found, is the corresponding machine name in the list of existing roles for the case type
       var names = _.pluck($scope.caseType.definition.caseRoles, 'name');
-      if (!_.contains(names, roleName)) {
-        var matchingRoles = _.filter($scope.relationshipTypeOptions, {id: roleName});
-        if (matchingRoles.length) {
-          var matchingRole = matchingRoles.shift();
-          roles.push({name: roleName, displaylabel: matchingRole.text});
-        } else {
-           CRM.loadForm(CRM.url('civicrm/admin/reltype', {action: 'add', reset: 1, label_a_b: roleName}))
-            .on('crmFormSuccess', function(e, data) {
-              var newType = _.values(data.relationshipType)[0];
-              roles.push({name: newType.label_b_a, displaylabel: newType.label_a_b});
-              // Assume that the case role should be A-B but add both directions as options.
-              $scope.relationshipTypeOptions.push({id: newType.label_a_b, text: newType.label_a_b});
-              if (newType.label_a_b != newType.label_b_a) {
-                $scope.relationshipTypeOptions.push({id: newType.label_b_a, text: newType.label_b_a});
-              }
-              $scope.$digest();
-            });
+      if (matchingRole) {
+        if (!_.contains(names, matchingRole.xmlNameOtherDirection)) {
+          roles.push({name: matchingRole.xmlNameOtherDirection, displaylabel: matchingRole.text});
```

Comment:
The `name` here (and in the lookup above) should be `matchingRole.xmlName`, since it's the value that would be going in the XML (from the non-client's perspective).  The `displaylabel` is correct here, since `matchingRole.text` will be what the user is seeing (from the client's perspective).